server/mqtt.go

```diff
@@ -2164,10 +2177,16 @@ func (as *mqttAccountSessionManager) removeSession(sess *mqttSession, lock bool)
 	}
 }
 
-// Helpers that sets the sub's mqtt fields and possibly serialize
-// (pre-loaded) retained messages.
-// Session lock held on entry.
+// Helpers that sets the sub's mqtt fields and possibly serialize (pre-loaded)
+// retained messages.
+//
+// Session lock held on entry. Acquires the session subs lock and holds it for
+// the duration. Non-MQTT messages coming into mqttDeliverMsgCbQoS0 will be
+// waiting.
 func (sess *mqttSession) processSub(c *client, subject, sid []byte, isReserved bool, qos byte, jsDurName string, h msgHandler, initShadow bool) (*subscription, error) {
+	sess.subsMu.Lock()
```

Comment:
This is the main fix. The only `sub.mqtt` data we need for QOS0 delivery is qos; so make a shorter-span `sess.subsMu` to protect it.

Reply:
Are you sure? We need to have retained messages prepped and in the `prm` writer prior to a callback be able to proceed. I don't think to recall that retained messages do not apply to QoS0 callback, at least I don't see that we don't do retained when qos==0 in `serializeRetainedMsgsForSub()`

Reply:
(resolving this in favor of the other thread)